finagle/h2/src/main/scala/com/twitter/finagle/buoyant/h2/BufferedStream.scala

```diff
@@ -96,9 +87,22 @@ class BufferedStream(underlying: Stream, bufferCapacity: Long = 8.kilobytes.byte
     }
   }
 
-  private[this] sealed trait PullState
-  private[this] case object Idle extends PullState
-  private[this] case class Pulling(f: Future[Unit]) extends PullState
+  private[this] sealed trait PullState {
+    def pullFrame(): Future[Frame]
+  }
+  private[this] case object Idle extends PullState {
+    override def pullFrame(): Future[Frame] = {
```

Comment:
Probably worth commenting that this is NOT thread-safe and must be externally synchronized (as it is in fork::read)